### ISSUE TITLE
Improve SOS so that it displays profiler modified IL for functions

### ISSUE THREAD

**Issue by noahfalk** [dotnet/coreclr#12719]
When jitting fails we'd like to know whether the JIT was operating on the original IL or profiler modified IL. We should identify one or more commands to change/modify (!DumpMD is probably one) to expose this information. In the runtime this info can be found in Module->m_debuggerSpecificData.m_pTemporaryILBlobTable
BaseDomain->m_rejitMgr->m_table (which is in progress of being refactored to CodeVersionManager)

We should also expose some basic information about whether or not a profiler is attached and what configuration flags it has set up. This info can be found in:
g_profControlBlock

**Comment by mmayr-at** [comment]
I was looking for a way to do this and tried a few things in the past few days until I've found this issue. 

First of all i think there are two cases we have to distinguish:
1. show profiler modified IL from existing functions
2. show IL code from functions, created via the IMetadataEmit interface (similar to #9827 / cc @discostu105)

For 1. I managed to access the profiler modified IL code via the `ICorDebug` interfaces. I've prototyped this with the DumpIL command. Here is the relevant source:

```
ToRelease<ICorDebugAppDomain> pCurrentAppDomain;
// get current appdomain ...
if (SUCCEEDED(pCurrentAppDomain->GetModuleFromMetaDataInterface(pImport, &pModule)) &&
	SUCCEEDED(pModule->GetFunctionFromToken(MethodDescData.MDToken, &pFunction)) &&
	SUCCEEDED(pFunction->GetILCode(&pCode)) &&
	SUCCEEDED(pCode->IsIL(&isIL)) && isIL)
{
	ULONG32 codeSize;
	ULONG32 fetched = 0;

	pCode->GetSize(&codeSize);
	
	BYTE *buffer = new BYTE[codeSize];
	pCode->GetCode(0, codeSize, codeSize, buffer, &fetched);
	
        // buffer contains raw IL code ...
}
```

Unfortunately the `GetCode` method returns only the raw IL code without the header, which makes decoding harder.

Here is the full DumpIL function: https://gist.github.com/mmayr-at/cebaaa17bba8347be471d283800e152d#file-dumpil-cpp-L89

For 2. I did not find a solution, because the `ICorDebugModule::GetFunctionFromToken()` fails to resolve the method token. This is the same behavior as other SOS commands, which do not work with newly emitted functions/types. Fixing this would be a great improvement for debugging profilers.

**Comment by noahfalk** [comment]
Check out ICorDebugProcess7::SetWriteableMetadataUpdateMode(). By default in a dump/snapshot debugging scenario ICorDebug exposes the original metadata for loaded modules because that is the behavior older debuggers expected and we didn't want to break back-compat. However if you call that API and specify AlwaysShowUpdates then a new code path will activate inside the metadata reader that should observe profiler updates.

As for grabbing the header, if you want to expose more data out of the ILCode object that seems reasonable. There are also some pre-existing APIs on ILCode such as GetEHClauses and GetLocalVarSigToken that expose some of the more commonly used data that is encoded in the header.

Hope that helps.

**Comment by mmayr-at** [comment]
This definitely helps, thank you.

When decoding the IL, I would like to resolve MemberRefs to see, what functions are called.  Unfortunately I haven't found a way to do this with the ICorDebug interface and the IMetadataEmit interface does not work with newly added functions. Do you know if this is possible?

**Comment by noahfalk** [comment]
IMetadataEmit is for creating new metadata and it sounds like you only want to read metadata that a profiler previously wrote. In that case you want to use IMetaDataImport which should be able to expose  profiler added references. Look at IMetadataImport::GetMemberRefProps.

**Comment by mmayr-at** [comment]
Sorry for the confusion, I actually meant the IMetadataImport interface. I already tested the  IMetadataImport::GetMemberRefProps method and it didn't work. But i will re-check this.

**Comment by noahfalk** [comment]
If you do confirm it wasn't working then it sounds like a bug. If you are up for it you could try to fix it yourself as part of your work here or feel free to file a separate issue on it.

**Comment by mmayr-at** [comment]
I can confirm that the IMetadataImport::GetMemberRefProps is also working for profiler generated member refs (at least for the desktop CLR). I've written a small WinDbg extension (will be published soon) to show the profiler generated IL code. I will integrate the results here, when I can find the time for it. 

I haven't managed to retrieve the EH clauses though. Looking at the [source](https://github.com/dotnet/coreclr/blob/1c9eb774950c98ae65ef5497d805cff2eb565971/src/debug/di/module.cpp#L3577) it looks like this information is only available when ReJIT is enabled, right?

**Comment by noahfalk** [comment]
> I haven't managed to retrieve the EH clauses though. Looking at the source it looks like this information is only available when ReJIT is enabled, right?

Yeah that appears correct. I had forgotten about it but it looks like we still have a gap if you are using the older SetILFunctionBody IL instrumentation mechanism at module load time (guessing that is your case?).

**Comment by mmayr-at** [comment]
> Yeah that appears correct. I had forgotten about it but it looks like we still have a gap if you are using the older SetILFunctionBody IL instrumentation mechanism at module load time (guessing that is your case?).

Yeah I am using the SetILFunctionBody mechanism in this case. Thanks for that information.

**Comment by sywhang** [comment]
> We should also expose some basic information about whether or not a profiler is attached and what configuration flags it has set up. This info can be found in:
> g_profControlBlock

FYI this part of the issue was moved to https://github.com/dotnet/coreclr/issues/18133 for the 6/2 Hackathon.
